muffnn/autoencoder/tests/test_autoencoder.py

```diff
@@ -0,0 +1,539 @@
+                                                          " shape!")
+
+    assert Xdec.shape == X.shape, ("Decoded iris data is not the right "
+                                   "shape!")
+
+    # One-hot encoded stuff should come back out normalized.
+    if cat_size is not None:
+        for begin, size in zip(cat_begin, cat_size):
+            assert_array_almost_equal(
+                np.sum(Xdec[:, begin: begin + size], axis=1), 1.0)
+
+    # Compute and test the scores.
+    scores = 0.0
+    for i in range(X.shape[1]):
+        if binary_inds is not None and i in binary_inds:
+            scores += sigmoid_cross_entropy(X[:, i:i+1], Xdec[:, i:i+1])
+        elif cat_size is not None and i in cat_begin:
+            ind = cat_begin.index(i)
+            b = cat_begin[ind]
+            s = cat_size[ind]
+            scores += softmax_cross_entropy(X[:, b:b+s], Xdec[:, b:b+s])
+        elif i in def_inds:
+            if metric == 'mse':
+                scores += np.sum((X[:, i:i+1] - Xdec[:, i:i+1]) ** 2, axis=1)
+            else:
+                scores += sigmoid_cross_entropy(X[:, i:i+1], Xdec[:, i:i+1])
+
+    ae_scores = ae.score_samples(X)
+    assert_array_almost_equal(scores, ae_scores, decimal=5)
+
+    score = np.mean(scores)
+    ae_score = ae.score(X)
+    assert_almost_equal(score, ae_score, decimal=5)
+
+    _LOGGER.warning("\ntest info:\n    ae: %s\n    sparse format: %s\n"
+                    "    score: %g\n    X[10]: %s\n    Xdec[10]: %s",
+                    str(ae), sparse_type, ae_score,
+                    pprint.pformat(list(X[10]), compact=True),
+                    pprint.pformat(list(Xdec[10]), compact=True))
+
+    assert ae_score < max_score, ("Autoencoder should have a score "
+                                  "less than %f for the iris features." %
+                                  max_score)
+
+    return ae_score
+
+
+MSE_MAX_SCORE = 0.1
+
+
+def test_mse_single_hidden_unit():
+    """Test the MSE metric w/ a single hidden unit."""
+    _check_ae(MSE_MAX_SCORE, hidden_units=(1,))
+
+
+def test_mse_multiple_hidden_units():
+    """Test the MSE metric w/ 2 hidden units."""
+    _check_ae(MSE_MAX_SCORE, hidden_units=(2,))
+
+
+def test_mse_multiple_layers():
+    """Test the MSE metric w/ layers (3, 2)."""
+    _check_ae(MSE_MAX_SCORE, hidden_units=(3, 2))
+
+
+def test_mse_dropout():
+    """Test the MSE metric w/ dropout."""
+    ae_score_dropout = _check_ae(MSE_MAX_SCORE,
+                                 hidden_units=(20, 20, 10, 10, 2),
+                                 dropout=0.05,
+                                 learning_rate=1e-3)
+
+    ae_score_nodropout = _check_ae(MSE_MAX_SCORE,
+                                   hidden_units=(20, 20, 10, 10, 2),
+                                   dropout=0.0,
+                                   learning_rate=1e-3)
+
+    assert ae_score_nodropout < ae_score_dropout, ("MSE with dropout should "
```

Comment:
Empirically I have never seen this happen. You do get better on models on *test* data, but this is just eliminating overfitting.